File: daily/101/solution.js
// 题目来源：https://leetcode-cn.com/problems/symmetric-tree/

var isSymmetric = function(root) {
    var stack = [];
    if(root === null) return true;
    if(root.left === null && root.right === null) return true;

    let left = null;
    let right = null;
    stack.push(root.right);
    stack.push(root.left);
    while(stack.length) {
        left = stack.pop();
        right = stack.pop();
        if(left === null || right === null || left.val !== right.val) {
            console.log(1);
            return false;
        }

        if((left.left !== null && right.right !== null) || (left.left === null && right.right === null)) {
            if(left.left !== null) { 
                stack.push(right.right);
                stack.push(left.left);
            }
        } else {
            console.log(2);
            return false;
        }

        if((left.right !== null && right.left !== null) || (left.right === null && right.left === null)) {
            if(left.right !== null) {
                stack.push(right.left);
                stack.push(left.right);
            }
        } else {
            console.log(3);
            return false;
        }
    }

    return true;
};

File: daily/101/solution.cpp
//recursion
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        return check(root->left, root->right);        
    }

    bool check(TreeNode *p, TreeNode *q){
        if(!p && !q) return true;
        if(!p || !q) return false;
        return p->val == q->val && check(p->left, q->right) && check(q->left, p->right);
    }
};

//iteration
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return true;
        return check(root->left, root->right);        
    }

    bool check(TreeNode *u, TreeNode *v){
        queue <TreeNode*> q;
        q.push(u); q.push(v);
        while(!q.empty()){
            u = q.front(); q.pop();
            v = q.front(); q.pop();
            if(!u && !v) continue;
            if((!u || !v) || (u->val != v->val)) return false;
            q.push(u->left);
            q.push(v->right);
            q.push(u->right);
            q.push(v->left);
        }
        return true;
    }
};